x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/loadingservice/LocalModel.java

```diff
@@ -53,23 +57,18 @@ public String getResultsType() {
                 return RegressionInferenceResults.NAME;
             default:
                 throw ExceptionsHelper.badRequestException("Model [{}] has unsupported target type [{}]",
-                    modelId,
-                    trainedModelDefinition.getTrainedModel().targetType());
+                        modelId,
+                        trainedModelDefinition.getTrainedModel().targetType());
         }
     }
 
     @Override
-    public void infer(Map<String, Object> fields, InferenceConfig config, ActionListener<InferenceResults> listener) {
-        try {
-            if (fieldNames.stream().allMatch(f -> MapHelper.dig(f, fields) == null)) {
-                listener.onResponse(new WarningInferenceResults(Messages.getMessage(INFERENCE_WARNING_ALL_FIELDS_MISSING, modelId)));
-                return;
-            }
-
-            listener.onResponse(trainedModelDefinition.infer(fields, config));
-        } catch (Exception e) {
-            listener.onFailure(e);
+    public InferenceResults infer(Map<String, Object> fields, InferenceConfig config) {
```

Comment:
I think this is backwards. 

We should have the synchronous method call the asynchronous method. This is the prevelant pattern everywhere else. Also, it is possible to make something asynchronous -> synchronous, not really the other way around.

Reply:
Yes it is backwards.

The function called by `LocalModel::infer` is `TrainedModelDefinition::infer` which does not have an async version. In this case we want to work to be done in the calling thread because the model is local to the call, for single threaded models I can't think of a situation where we would want to spawn another thread to do the work as we know inference is cheap. For models that could be parallelised and the work split over multiple threads then yes you would want to make it async. 

Do we even need the async method right now? It is only called by [TransportInternalInferModelAction](https://github.com/elastic/elasticsearch/blob/a629efebc503b8db34ae3c38007fed7a7f3a4498/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportInternalInferModelAction.java#L65) and could easily be changed.

I removed the default method because it is backwards and wrong then implemented it in `LocalModel`.

Reply:
> Do we even need the async method right now? 

Maybe not, but it is much more difficult to make things asynchronous once they are synchronous. 

Assumptions are made about where the model lives when it is synchronous.

What if this was a natively loaded model? 
Would we pause the calling thread for the data to be serialized down the native process?

I am not sure about this, but I did not want to paint us in a corner.